### ISSUE TITLE
VSL Policy Collapse Issue: Speed Limits Over-Converge to a Straight Line

### ISSUE THREAD

**Issue by via-life** [juanwulu/sac_vsl#2]
@juanwulu Hi,

Thank you for your patient guidance in our previous email communication. Following your suggestion, I am summarizing the issue here for easier tracking and discussion.

Problem Description:
When reproducing the SAC_VSL project, the speed limits output by my trained model over-converge, appearing almost as a straight line, which differs significantly from the results shown in your paper/example.

What I have tried:

Dependency versions: Checked software package versions against official documentation and confirmed they are consistent with the original project. Ruled out version differences as the primary cause. Environment details:

OS: Windows 11

Python: 3.11

Key packages: gputil==1.4.0, gym==0.23.1, gymnasium==0.26.2, lz4==4.3.2, numpy==1.24.3, scipy==1.15.2, tensorboard==2.12.3, tensorboardX==2.6.4, ray==2.6.3, ray[rllib]==2.6.3, ray[tune]==2.6.3

Reward penalty term: Tried adjusting the penalty factor (removed + 0.2 * penalty from the original reward), but the issue persists.

Buffer size: Due to local hardware constraints, changed --buffer-size from 1e6 to 2e5. I believe this is unlikely to be the main cause.

<img width="5435" height="2963" alt="Image" src="https://github.com/user-attachments/assets/b6297c46-12d0-48b7-ad16-dec81fbb8326" />

[sac_vsl_1.2 - 20260401.zip](https://github.com/user-attachments/files/26478983/sac_vsl_1.2.-.20260401.zip)

**Comment by juanwulu** [comment]
@via-life Hi there,

I managed to reproduce my results in the report on my local machine, and here are my specs for your reference:

```text
OS: Ubuntu 24.04.4 LTS x86_64
Kernel: Linux 6.8.0-106-generic
Uptime: 4 days, 11 hours, 5 mins
Packages: 4438 (dpkg), 10 (flatpak), 34 (snap)
Shell: zsh 5.9
Display (DELL P2222H): 1920x1080 in 22", 60 Hz [External]
DE: GNOME 46.0
Terminal: tmux 3.4
CPU: 12th Gen Intel(R) Core(TM) i9-12900KF (24) @ 5.20 GHz
GPU: NVIDIA GeForce RTX 3090 Ti [Discrete]
Memory: 7.85 GiB / 62.59 GiB (13%)
Swap: 5.48 GiB / 7.63 GiB (72%)
```
---

I notice several discrepancies between your implementation and my reproduction:

| | Your Setup | Original Project |
|:---|:---|:---|
| Python | 3.11 | 3.10 |
| Ray/RLlib | 2.6.3 | 2.1.0 |
| Gym | 0.23.1 + gymnasium 0.26.2 | 0.23.1 only |

The Ray version difference is the most damaging. Between 2.1.0 and 2.6.3, RLlib migrated to an entirely new internal API (`RLModule`). When the training script passes the legacy config dictionary (e.g., `q_model_config`, `policy_model_config`, global `lr`) into Ray 2.6, these keys are **silently ignored** — no error, no warning. Concretely, this means:

- The `[256, 256]` hidden layer architecture you requested via `q_model_config` and `policy_model_config` is **dropped**. The agent trains with a default architecture that doesn't match the observation space.
- The global `lr: 0.0025` is **ignored**. The agent falls back to the internal default of `0.0003`.
- `rollout_fragment_length` changed from `1` (update after every environment step) to `"auto"` (much larger fragments), which drastically slows down off-policy learning and starves early training.

Please compare the SAC implementation in `ray==2.6.3` [here](https://github.com/ray-project/ray/blob/ray-2.6.3/rllib/algorithms/sac/sac.py) and the one in `ray==2.1.0` [here](https://github.com/ray-project/ray/blob/ray-2.1.0/rllib/algorithms/sac/sac.py) for your reference.

Having both `gym==0.23.1` and `gymnasium==0.26.2` installed simultaneously also causes issues — Ray 2.6 defaults to the `gymnasium` API, which expects `step()` to return `(obs, reward, terminated, truncated, info)` instead of the `(obs, reward, done, info)` tuple this environment produces.

The combination of a wrong network architecture, a wrong learning rate, a wrong update frequency, and a mismatched environment API is enough on its own to guarantee immediate policy collapse.

**In summary, to reproduce the original results, please use:**
- Python **3.10.x** or **3.9.x**
- `ray==2.1.0`
- `gym==0.23.1` (and **remove** `gymnasium` if installed)
- SUMO **1.15.0** if possible (the project was developed against this version; newer versions may shift simulation dynamics)

**Comment by juanwulu** [comment]
> In the past two weeks, I've been trying to narrow down the issue. A couple of days ago, I suspected that the problem might be related to the tt calculation region in the reward function — which was originally located downstream of the VSL zone. This could have led the model to learn a strategy that actively restricts upstream traffic flow, so that vehicles entering the tt calculation region can pass more quickly.
> 
> To test this, I expanded the tt region to "200m downstream of the VSL zone + the original 200m region". However, after retraining, the model still exhibits the same problematic behavior as before.
> 
> After two weeks of systematically ruling out other possibilities, I now suspect that the main issue lies in the reward function. I noticed that some reward factors are commented out in env.py. Could this be related to the problem I'm encountering? I would really appreciate your insight.

Thanks for the great insights, @via-life! My apologies for the late response. After reviewing the documentation, I find that my previous implementation of a reward based on travel time may introduce a loophole. Please check #5 for more details.

**Comment by juanwulu** [comment]
Close this issue due to inactivity.
